test/extended/apiserver/tls.go

```diff
@@ -1,31 +1,118 @@
 package apiserver
 
 import (
+	"context"
 	"crypto/tls"
+	"fmt"
+	"strings"
+	"time"
 
 	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	"github.com/openshift/library-go/test/library"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 
+	operatorv1client "github.com/openshift/client-go/operator/clientset/versioned"
 	"github.com/openshift/library-go/pkg/crypto"
 	exutil "github.com/openshift/origin/test/extended/util"
 )
 
+const tlsTestDuration = 45 * time.Minute
+const tlsWaitForCleanupDuration = 10 * time.Minute
+
+var _ = g.Describe("[sig-api-machinery][Feature:APIServer][Serial]", func() {
+	defer g.GinkgoRecover()
+
+	oc := exutil.NewCLIWithoutNamespace("apiserver")
+
+	g.It("TestTLSModernProfile", func() {
+		ctx, ctxCancelFn := context.WithTimeout(context.Background(), tlsTestDuration)
+		defer ctxCancelFn()
+
+		t := g.GinkgoT()
+
+		configClient := oc.AdminConfigClient()
+
+		operatorClient := oc.AdminOperatorClient()
+
+		defer func() {
+			g.By("Cleanup - removing TLS profile")
+
+			cleanupCtx, cancelCleanup := context.WithTimeout(context.Background(), tlsWaitForCleanupDuration)
+			defer cancelCleanup()
+
+			kasStatus, err := operatorClient.OperatorV1().KubeAPIServers().Get(cleanupCtx, "cluster", metav1.GetOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			currentRevision := kasStatus.Status.LatestAvailableRevision
+
+			removeModernProfilePatch := `[{"op":"remove","path":"/spec/tlsSecurityProfile"}]`
+
+			_, err = configClient.ConfigV1().APIServers().Patch(cleanupCtx, "cluster", types.JSONPatchType, []byte(removeModernProfilePatch), metav1.PatchOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			_, err = waitForKASIncrementedRevision(cleanupCtx, operatorClient, "cluster", currentRevision)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}()
+
+		g.By("Applying a JSON Patch to use the modern TLS profile")
+
+		kasStatus, err := operatorClient.OperatorV1().KubeAPIServers().Get(ctx, "cluster", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		currentKASRevision := kasStatus.Status.LatestAvailableRevision
+
+		addModernProfilePatch := `[{"op":"add","path":"/spec/tlsSecurityProfile","value":{"type":"Modern","modern":{}}}]`
+
+		_, err = configClient.ConfigV1().APIServers().Patch(ctx, "cluster", types.JSONPatchType, []byte(addModernProfilePatch), metav1.PatchOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("Waiting for etcd to stabilize")
+
+		podClient := oc.AdminKubeClient().CoreV1().Pods("openshift-etcd")
+		err = library.WaitForPodsToStabilizeOnTheSameRevision(t, podClient, "app=etcd", 5, 24*time.Second, 5*time.Second, 30*time.Minute)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("Waiting for the API server to stabilize")
+
+		_, err = waitForKASIncrementedRevision(ctx, operatorClient, "cluster", currentKASRevision)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("Dialing the API with a minimum TLS version of 1.3")
```

Comment:
shouldn't we check if dialing with a version < 1.3 fails ?

Reply:
agreed - updated